File: nr02/rational.cpp

#include "rational.h"

// Complete these methods:

int rational::gcd( int n1, int n2 ) {
  while (n2) {
    n1 %= n2;
    std::swap(n1, n2);
  }
  return n1;
}

void rational::normalize( ) {
  int g = rational::gcd(num, denum);

  num /= g;
  denum /= g;
  if (denum < 0) {
    denum *= -1;
    num *= -1;
  }
}

rational operator - ( rational r ) {
  rational tmp(-r.num, r.denum);
  return tmp;
}

rational operator + ( const rational& r1, const rational& r2 ) {
  int n_num = r1.num * r2.denum + r1.denum * r2.num;
  int n_denum = r1.denum * r2.denum;
  rational tmp(n_num, n_denum);
  return tmp;
}

rational operator - ( const rational& r1, const rational& r2 ) {
  return r1 + (-r2);
}

rational operator * ( const rational& r1, const rational& r2 ) {
  int n_num = r1.num * r2.num;
  int n_denum = r1.denum * r2.denum;
  rational tmp(n_num, n_denum);
  return tmp;
}

rational operator / ( const rational& r1, const rational& r2 ) {
    int n_num = r1.num * r2.denum;
    int n_denum = r1.denum * r2.num;
    rational tmp(n_num, n_denum);
    return tmp;
}

bool operator == ( const rational& r1, const rational& r2 );
bool operator != ( const rational& r1, const rational& r2 );

std::ostream& operator << ( std::ostream& out, const rational& r ) {
  if (r.denum == 1)
    out << r.num;
  else
    out << r.num << "/" << r.denum;
  return out;
}

File: python/3/rational.py
from numbers import Number
from vector import *
from matrix import *

class Rational( Number ) :

	def __init__( self, num, denum = 1 ) :
		self.num = num
		self.denum = denum
		self.normalize()
	def normalize(self):
		if self.denum < 0:
			self.denum *= -1
			self.num *= -1
		if self.denum == 0:
			raise ArithmeticError( "Denum can not be zero" )
		n = gcd(self.num, self.denum)
		if n!=1 and n!=-1:
			self.num //= n
			self.denum //= n

	def __repr__(self):
		return str(self.num) + " / " + str(self.denum)
	
	def __neg__(self):
		return Rational(self.num * -1, self.denum)
	
	def __add__(self, other):
		if isinstance(other, int):
		    other = Rational(other)
		num = self.num * other.denum + other.num * self.denum
		denum = self.denum * other.denum
		return Rational(num, denum)

	def __sub__(self, other):
		if isinstance(other, int):
		    other = Rational(other)
		num = self.num * other.denum - other.num * self.denum
		denum = self.denum * other.denum
		return Rational(num, denum)
			

	def __radd_(self, other):
		return self + other

	def __rsub_(self, other):
		return -(self - other)
	
	def __mul__(self, other):
		if isinstance(other, int):
		    other = Rational(other)
		num = self.num * other.num
		denum = self.denum * other.denum
		return Rational(num, denum)

	def __truediv__(self, other):
		if isinstance(other, int):
		    other = Rational(other)
		num = self.num * other.denum
		denum = self.denum * other.num
		return Rational(num, denum)

	def __rmul__(self, other):
		return self * other

	def __rtruediv__(self, other):
		tmp = self / other
		return Rational(tmp.denum, tmp.num)
		
	
		
		
		

def gcd(n1, n2):
	if n1==0 and n2==0:
		raise ArithmeticError( "gcd(0,0) does not exist" )	
	while n2:
		n1, n2 = n2, n1%n2
	return n1

print(Rational(-7, 28))
